pkg/R/DataFrame.R

```diff
@@ -114,20 +114,192 @@ setMethod("count",
             callJMethod(sdf, "count")
           })
 
-#' Collect elements of a DataFrame
-#' 
-#' Returns a list of Row objects from a DataFrame
-#' 
-#' @param df A SparkSQL DataFrame
-#' 
+# Collect the rows of a Spark DataFrame and return a named list for each row.
+
 #' @rdname collect-methods
 #' @export
+#' @examples
+#'\dontrun{
+#' sc <- sparkR.init()
+#' sqlCtx <- sparkRSQL.init(sc)
+#' path <- "path/to/file.json"
+#' df <- jsonFile(sqlCtx, path)
+#' collected <- collect(df)
+#' firstName <- collected[[1]]$name
+#' }
 
-# TODO: Collect() currently returns a list of Generic Row objects and is WIP.  This will eventually 
-# be part of the process to read a DataFrame into R and create a data.frame.
 setMethod("collect",
           signature(rdd = "DataFrame"),
-          function(rdd){
-            sdf <- rdd@sdf
-            listObj <- callJMethod(sdf, "collect")
+          function(rdd) {
+            rddIn <- toRDD(rdd)
+            collect(rddIn)
+          })
+
+# Take the first NUM elements in a DataFrame and return a named list for each row.
+
+#' @rdname take
+#' @export
+#' @examples
+#'\dontrun{
+#' sc <- sparkR.init()
+#' sqlCtx <- sparkRSQL.init(sc)
+#' path <- "path/to/file.json"
+#' df <- jsonFile(sqlCtx, path)
+#' take(df, 2)
+#' }
+
+setMethod("take",
+          signature(rdd = "DataFrame", num = "numeric"),
+          function(rdd, num) {
+            rddIn <- toRDD(rdd)
+            take(rddIn, num)
+          })
+
+#' toRDD()
+#' 
+#' Converts a Spark DataFrame to an RDD while preserving column names.
+#' 
+#' @param df A Spark DataFrame
+#' 
+#' @rdname DataFrame
+#' @export
+#' @examples
+#'\dontrun{
+#' sc <- sparkR.init()
+#' sqlCtx <- sparkRSQL.init(sc)
+#' path <- "path/to/file.json"
+#' df <- jsonFile(sqlCtx, path)
+#' rdd <- toRDD(df)
+#' }
+
+setGeneric("toRDD", function(df) { standardGeneric("toRDD") })
+
+setMethod("toRDD",
+          signature(df = "DataFrame"),
+          function(df) {
+            jrdd <- callJStatic("edu.berkeley.cs.amplab.sparkr.SQLUtils", "dfToRowRDD", df@sdf)
+            names <- callJStatic("edu.berkeley.cs.amplab.sparkr.SQLUtils", "getColNames", df@sdf)
+            RDD(jrdd, serializedMode = "row", colNames = names)
```

Comment:
I'm not sure it's quite that easy. The `colNames` slot was actually added because I needed a way to pass the column names into `RRDD.scala` and, subsequently, to the R workers when evaluating UDFs from the R side. By putting them into the `colNames` slot, we ensure they have a placeholder when it comes time to deserialize all of the information on the R worker side after the `RRDD` constructor has been called. This also factors into PipelinedRDDs and being able to perform multiple transformation on an RDD in `row` mode.

I'm not 100% sure either way and I think it's worth discussing more, but I do know that there's more to it than just needing to lapply names onto an RDD after the fact.

What do you guys think?